search-service/src/main/kotlin/com/egm/stellio/search/web/TemporalApiResponseBuilder.kt

```diff
@@ -0,0 +1,241 @@
+                TemporalQuery.Timerel.BEFORE -> temporalQuery.timeAt!!
+                TemporalQuery.Timerel.BETWEEN -> temporalQuery.endTimeAt!!
+                else -> discriminatingTimeRange.first
+            }
+            return rangeStart to discriminatingTimeRange.second
+        } else {
+            val discriminatingTimeRange = attributesTimeRanges.minOf { it.first } to
+                attributesTimeRanges.minOf { it.second }
+            val rangeStart = when (temporalQuery.timerel) {
+                TemporalQuery.Timerel.AFTER -> temporalQuery.timeAt!!
+                TemporalQuery.Timerel.BETWEEN -> temporalQuery.timeAt!!
+                else -> discriminatingTimeRange.first
+            }
+            return rangeStart to discriminatingTimeRange.second
+        }
+    }
+
+    private fun filterEntityInRange(
+        entity: CompactedEntity,
+        range: Range,
+        query: TemporalEntitiesQuery
+    ): CompactedEntity {
+        return entity.keys.mapNotNull { key ->
+            when (val attribute = entity[key]) {
+                is List<*> -> {
+                    val temporalAttribute = attribute as CompactedTemporalAttribute
+                    // faster filter possible because list is already ordered
+                    val filteredAttribute = temporalAttribute.filter { range.contain(it, query.temporalQuery) }
+                    if (filteredAttribute.isNotEmpty()) key to filteredAttribute else key to emptyList()
+                }
+
+                is Map<*, *> -> {
+                    attribute.toTemporalValuesOrNull(query)?.let {
+                        return@mapNotNull key to it.filter { range.contain(it) }
+                    }
+                    if (range.contain(attribute as CompactedAttribute, query.temporalQuery)) key to attribute
+                    else key to emptyList<CompactedAttribute>()
+                }
+
+                else -> key to (attribute ?: return@mapNotNull null)
+            }
+        }.toMap()
+    }
+
+    private fun Range.contain(time: ZonedDateTime): Boolean =
+        (this.first >= time && time >= this.second) || (this.first <= time && time <= this.second)
+
+    private fun Range.contain(temporalValue: TemporalValue): Boolean =
+        this.contain(ZonedDateTime.parse(temporalValue[1] as String))
+
+    private fun Range.contain(attribute: CompactedAttribute, query: TemporalQuery): Boolean =
+        attribute.getAttributeDate(query) ?.let { this.contain(it) } ?: true // if no date it should not be filtered
+
+    private fun CompactedAttribute.getAttributeDate(query: TemporalQuery): ZonedDateTime? =
+        this[query.timeproperty.propertyName]?.let {
+            ZonedDateTime.parse(it as String)
+        }
+
+    private fun Map<*, *>.toTemporalValuesOrNull(query: TemporalEntitiesQuery): List<TemporalValue>? {
+        if (!query.withTemporalValues) null
+        SIMPLIFIED_TEMPORAL_VALUE_NAMES.forEach { name ->
```

Comment:
instead of having to maintain a list, you could say "anything that is not `type`, nor `datasetId`"

Reply:
Hum, yes but this mean maintaining a list of non temporalValues possible keys. 

Is there a chance of a new key of this type appearing. if not i will change the code.

If there is a chance of a new key appearing i think its better to keep the code linked to the temporal_values_name since any change to them already implied we touch the temporal pagination implementation.